driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java

```diff
@@ -102,37 +107,43 @@
             this.operationTime = result.getTimestamp(OPERATION_TIME, null);
             this.postBatchResumeToken = getPostBatchResumeTokenFromResponse(result);
         }
-        if (firstQueryResult.getCursor() != null) {
+        ServerCursor serverCursor = initFromQueryResult(firstQueryResult);
+        if (serverCursor != null) {
             notNull("connectionSource", connectionSource);
         }
-        if (connectionSource != null) {
-            this.connectionSource = connectionSource.retain();
-        }
-
-        initFromQueryResult(firstQueryResult);
         firstBatchEmpty = firstQueryResult.getResults().isEmpty();
-
+        Connection connectionToPin = null;
+        boolean releaseServerAndResources = false;
         if (connection != null) {
             this.maxWireVersion = connection.getDescription().getMaxWireVersion();
             if (limitReached()) {
-                killCursor(connection);
+                releaseServerAndResources = true;
             } else {
                 assertNotNull(connectionSource);
                 if (connectionSource.getServerDescription().getType() == ServerType.LOAD_BALANCER) {
-                    this.connection = connection.retain();
-                    this.connection.markAsPinned(Connection.PinningMode.CURSOR);
+                    connectionToPin = connection;
                 }
             }
         }
-        releaseConnectionAndSourceIfNoServerCursor();
+        closeableState = new CloseableState(namespace, connectionSource, connectionToPin, serverCursor);
+        if (releaseServerAndResources) {
+            closeableState.releaseServerAndClientResources(assertNotNull(connection));
+        }
     }
 
     @Override
     public boolean hasNext() {
-        if (closed) {
+        if (!closeableState.startOperation()) {
             throw new IllegalStateException("Cursor has been closed");
         }
+        try {
+            return doHasNext();
+        } finally {
+            closeableState.endOperation();
```

Comment:
This comment elaborates on  `startOperation`/`endOperation` and the extracted `do...` methods.

`QueryBatchCursor` has five public methods that access releasable resources:

- `close`
- `getServerCursor`
- `hasNext`
- `next`
- `tryNext`

`getServerCursor` is a special trivial case that reads a reference to a releasable resource and returns it, so its implementation is trivial: it can either see the resource (if it was there in the first place and it was not killed) or not. We are not going to discuss this method anymore.

The rest four methods are less trivial. Methods `hasNext`, `next`, `tryNext` are supposed to be called sequentially because `QueryBatchCursor` is not thread-safe, but the method `close` can be called concurrently with either of them and with itself. We want all the releasable resources to stay available until `hasNext`/`next`/`tryNext` is running, as otherwise they would have to be prepared to lose any resource at any point, which significantly complicates matters. Any of the three aforementioned operations signify that it is running by executing `startOperation`, and signifies that it stopped running `endOperation`.

The demarkation described above is complicated by the fact that operations may have mutual calls, e.g., `next` calls `hasNext`. This means that if `hasNext` signifies that it has completed, we cannot assume that the resources are not used anymore as the `next` that called `hasNext` may still be running. Surely this can be handled by counting the number of active operations, but a simpler approach is to get rid of the root cause: one operation calling another. This is what I did in this PR: the actual logic of the methods that use `startOperation`/`endOperation` was extracted to methods with the prefix `do` (`hasNext` -> `doHasNext`) where needed, such that the `do` methods can continue calling each other, while we can rely on resources not being used after `endOperation` returning and before `startOperation` starting. The `startOperation`/`endOperation` operations are now totally ordered with `start` always followed by `end`. These methods now enforce this order internally, which also allows us to detect invalid concurrent usage of the `hasNext`/`next`/`tryNext` methods.

If the resources are released after `endOperation` returning and before `startOperation` starting (which is the only logical time at which they could be released by the deferred close action), then the next `startOperation` returns `false`, signifying that the `QueryBatchCursor` has been closed.

Reply:
I think you can clean and encapsulate this pattern to inside the closeable state class:

Something like this (my naming isn't great):
```
        <A> A inOperation(final Supplier<A> supplier) {
            if (!startOperation()) {
                throw new IllegalStateException("Iterator has been closed");
            }
            try {
                return supplier.get();
            } finally {
                endOperation();
            }
        }
```

would allow for:

```
    @Override
    public boolean hasNext() {
        return closeableState.inOperation(this::doHasNext);
    }
    
    public List<T> next() {
        return closeableState.inOperation(this::doNext);
    }

```

I think something like that clarifies the intent more.